File: LtfsServer/API/LocalFileSystem.cs
using LtfsServer.Services;

namespace LtfsServer.API;

public static class APILocalFileSystem
{
    public static void MapLocalFileSystemApi(this WebApplication app)
    {
        app.MapGet("/api/fsroots", async (ILocalFileSystemTreeService service, HttpContext context) =>
        {
            var roots = await service.GetRootsAsync(context.RequestAborted);
            return Results.Ok(new
            {
                items = roots.Select(ToDto).ToArray(),
                loadedAtUtc = DateTime.UtcNow,
            });
        });

        app.MapGet("/api/fschildren", async (string? path, ILocalFileSystemTreeService service, HttpContext context) =>
        {
            if (string.IsNullOrWhiteSpace(path))
            {
                return Results.BadRequest(new { error = "Query string 'path' is required." });
            }

            try
            {
                var result = await service.GetChildrenAsync(path, context.RequestAborted);
                return Results.Ok(new
                {
                    parentPath = result.ParentPath,
                    items = result.Children.Select(ToDto).ToArray(),
                    warning = result.Warning,
                    loadedAtUtc = DateTime.UtcNow,
                });
            }
            catch (ArgumentException ex)
            {
                return Results.BadRequest(new { error = ex.Message });
            }
            catch (DirectoryNotFoundException ex)
            {
                return Results.NotFound(new { error = ex.Message });
            }
        });
    }

    private static object ToDto(LocalFsNode node)
    {
        return new
        {
            id = node.Id,
            name = node.Name,
            path = node.Path,
            kind = node.Kind,
            available = node.Available,
            hasChildren = node.HasChildren,
            error = node.Error,
        };
    }
}

File: ltfs-capybara-gui/src/api/modules/localfilesystem.ts
import { apiClient } from '../client';

export interface FsTreeNodeDto {
    id: string;
    name: string;
    path: string;
    kind: 'drive' | 'network' | 'dir';
    available: boolean;
    hasChildren: boolean;
    error?: string | null;
}

export interface FsRootsResponse {
    items: FsTreeNodeDto[];
    loadedAtUtc: string;
}

export interface FsChildrenResponse {
    parentPath: string;
    items: FsTreeNodeDto[];
    warning?: string | null;
    loadedAtUtc: string;
}

export const localFileSystemApi = {
    async getRoots(): Promise<FsRootsResponse | null> {
        try {
            const res = await apiClient.get<FsRootsResponse>('/fsroots');
            return res.data;
        } catch (err) {
            console.error('localFileSystemApi.getRoots error', err);
            return null;
        }
    },

    async getChildren(path: string): Promise<FsChildrenResponse | null> {
        try {
            const res = await apiClient.get<FsChildrenResponse>('/fschildren', {
                params: { path },
            });
            return res.data;
        } catch (err) {
            console.error('localFileSystemApi.getChildren error', err);
            return null;
        }
    },
};
